### PR TITLE
Feedback resource

### PR DESCRIPTION
**Pull request by RonAzar**
Complete Phase 1- Step 4.2

### CODE REVIEW

File: src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import org.springframework.stereotype.Controller
+
+@Controller
+@Path("/api/feedback")
+class FeedbackResource(private val feedbackService: FeedbackService) {
+    @Path("/submit")
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    fun submitFeedback(@Context requestContext: ContainerRequestContext, newFeedback: FeedbackRequest): Response {
+        val companyId = requestContext.getProperty("companyId") as? Number
```
Comment (vasofi) [inline src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt:21]: DRY - this logic of extracting data from the context repeats itself in all endpoints.
Try to refactor this

File: src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            ?: return Response.status(Response.Status.BAD_REQUEST).entity("Bad Request: Missing companyId").build()
```
Comment (vasofi) [inline src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt:22]: In this case, if there's a problem with the cookie - i would return NOT AUTHORIZED or NO PERMISSIONS, not bad request

Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt:22]: Got it
Fixed

File: src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt (hunk 1 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt:21]: Add two functions:
extractClaimAsLong, extractRole

File: src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        val roleString = requestContext.getProperty("role") as? String
+            ?: return Response.status(Response.Status.BAD_REQUEST).entity("Bad Request: Missing role").build()
+
+        val employeeRole = try {
+            EmployeeRole.valueOf(roleString.uppercase())
+        } catch (e: IllegalArgumentException) {
+            return Response.status(Response.Status.BAD_REQUEST).entity("Bad Request: Role not found!").build()
+        }
+
+        if (employeeRole == EmployeeRole.ADMIN || employeeRole == EmployeeRole.HR) {
```
[the remark this answers is not in the document]
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt]: provided with a filter mechanism inside RolePermissionValidator class

File: src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt (hunk 1 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt]: created new class called RolePermissionValidator with all the functionality

File: src/main/kotlin/com/hibob/academy/employee_feedback_feature/validation/RolePermissionValidator.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+            } ?: run {
+                requestContext.abortWith(
+                    Response.status(Response.Status.UNAUTHORIZED).entity("UNAUTHORIZED: Missing role").build()
+                )
+                null
+            }
+        }
+
+        fun hrOrAdminValidation(role: EmployeeRole): Boolean {
+            return role == EmployeeRole.ADMIN || role == EmployeeRole.HR
```
Comment (vasofi) [inline src/main/kotlin/com/hibob/academy/employee_feedback_feature/validation/RolePermissionValidator.kt:33]: We don't want this mechanism to be aware of why kinds of api endpoints we have, and by the names and validations that you have done here, it is very aware of it.

Also, think what happens if we need a new kind of validation, for example - isManager. It means that I have to come to this file and create a new method?

File: src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+@Controller
+@Path("/api/feedback")
+class FeedbackResource(private val feedbackService: FeedbackService) {
+    @Path("/submit")
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    fun submitFeedback(@Context requestContext: ContainerRequestContext, newFeedback: FeedbackRequest): Response {
+        val companyId = extractClaimAsLong(requestContext, "companyId")!!
+        val employeeId = if (!newFeedback.isAnonymous){
```
[the remark this answers is not in the document]
Comment (vasofi) [reply src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt:24]: This if is not very readable. Better to put the assignment of the variable inside the if

[the remark this answers is not in the document]
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt:24]: Got it!
remove space before brackets ( and {
but don't we prefer using val and not var?

File: src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+@Controller
+@Path("/api/feedback")
+class FeedbackResource(private val feedbackService: FeedbackService) {
+    @Path("/submit")
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    fun submitFeedback(@Context requestContext: ContainerRequestContext, newFeedback: FeedbackRequest): Response{
+        val companyId = extractClaimAsLong(requestContext, "companyId")!!
+        val employeeId = if(!newFeedback.isAnonymous){
```
Comment (vasofi) [inline src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt:25]: space before brackets in if

Comment (vasofi) [reply src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/FeedbackResource.kt:25]: this if is not very readable

**Review by vasofi — APPROVED** [review]

**Pull request merged**